File: binarysearch/Matrix-Prefix-Sum/Matrix-Prefix-Sum.cpp
// https://helloacm.com/compute-the-matrix-prefix-sum/
// https://binarysearch.com/problems/Matrix-Prefix-Sum
// EASY, PREFIX SUM, MATH

vector<vector<int>> solve(vector<vector<int>>& matrix) {
    if (matrix.empty()) return matrix;
    int rows = matrix.size();
    int cols = matrix[0].size();
    for (int r = 0; r < rows; ++ r) {
        for (int c = 1; c < cols; ++ c) {
            matrix[r][c] += matrix[r][c - 1];
        }
    }
    for (int r = 1; r < rows; ++ r) {
        for (int c = 0; c < cols; ++ c) {
            matrix[r][c] += matrix[r - 1][c];
        }
    }
    return matrix;
}

File: binarysearch/Matrix-Prefix-Sum/Matrix-Prefix-Sum.py
# https://helloacm.com/teaching-kids-programming-matrix-prefix-sum-algorithm/
# https://binarysearch.com/problems/Matrix-Prefix-Sum
# EASY, MATH

class Solution:
    def solve(self, matrix):
        if not matrix:
            return []

        row = len(matrix)
        if 0 == row:
            return matrix
        col = len(matrix[0])
        if 0 == col:
            return matrix

        for r in range(row):
            matrix[r] = list(accumulate(matrix[r]))

        for r in range(1, row):
            for c in range(col):            
                matrix[r][c] += matrix[r - 1][c]

        return matrix
